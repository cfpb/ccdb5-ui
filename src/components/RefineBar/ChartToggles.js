import './ChartToggles.less';
import { changeChartType } from '../../actions/trends';
import getIcon from '../iconMap';
import React from 'react';
import { selectedClass, sendAnalyticsEvent } from '../../utils';
import { useDispatch, useSelector } from 'react-redux';
import { selectTrendsChartType } from '../../reducers/trends/selectors';

export const ChartToggles = () => {
  const dispatch = useDispatch();
  const chartType = useSelector(selectTrendsChartType);

  const toggleChartType = (chartType) => {
    sendAnalyticsEvent('Button', 'Trends:' + chartType);
    dispatch(changeChartType(chartType));
  };

  return (
    <section className="chart-toggles m-btn-group">
      <p>Chart type</p>
      <button
        aria-label="Toggle line chart"
        className={'a-btn' + selectedClass('line', chartType)}
        disabled={chartType === 'line'}
        onClick={() => {
          toggleChartType('line');
        }}
      >
        {getIcon('line-chart')}
      </button>
      <button
        aria-label="Toggle area chart"
        className={'a-btn' + selectedClass('area', chartType)}
        disabled={chartType === 'area'}
        onClick={() => {
          toggleChartType('area');
        }}
      >
        {getIcon('area-chart')}
      </button>
    </section>
  );
};
