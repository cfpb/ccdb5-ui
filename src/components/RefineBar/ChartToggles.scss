.chart-toggles {
  button {
    width: 38px;
    background-color: var(--pacific-40);
    padding: 3px;
    svg {
      height: 25px;

      &#line-chart-icon {
        .cls-1 {
          opacity: 0.4;
        }

        .cls-2 {
          fill: none;
          stroke: #101820;
          stroke-miterlimit: 10;
          stroke-width: 2px;
        }

        .cls-3 {
          fill: #101820;
        }
      }
      &#area-chart-icon {
        .cls-1 {
          fill: #101820;
        }

        .cls-2 {
          opacity: 0.4;
        }
      }
    }
  }
}
