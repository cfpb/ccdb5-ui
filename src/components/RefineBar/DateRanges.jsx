import './DateRanges.less';
import { connect } from 'react-redux';
import { dateRanges } from '../../constants';
import { dateRangeToggled } from '../../actions/filter';
import React from 'react';
import { sendAnalyticsEvent } from '../../utils';

export class DateRanges extends React.Component {
  _btnClassName( dateRange ) {
    const classes = [ 'a-btn', 'date-selector', 'range-' + dateRange ];
    if ( dateRange === this.props.dateRange ) {
      classes.push( 'selected' );
    }
    return classes.join( ' ' );
  }

  _toggleDateRange( dateRange ) {
    if ( this.props.dateRange !== dateRange ) {
      this.props.toggleDateRange( dateRange, this.props.tab );
    }
  }

  render() {
    return (
      <section className="date-ranges m-btn-group">
        <p>Date range (Click to modify range)</p>
        { dateRanges.map( dateRange => <button onClick={ () => {
          this._toggleDateRange( dateRange );
        } }
        className={ this._btnClassName( dateRange ) }
        key={ dateRange }>
          { dateRange }
        </button>
        ) }
      </section>
    );
  }
}

export const mapStateToProps = state => ( {
  dateRange: state.query.dateRange,
  tab: state.query.tab
} );

export const mapDispatchToProps = dispatch => ( {
  toggleDateRange: ( range, tab ) => {
    sendAnalyticsEvent( 'Button', tab + ':' + range );
    dispatch( dateRangeToggled( range ) );
  }
} );

export default connect( mapStateToProps, mapDispatchToProps )( DateRanges );
