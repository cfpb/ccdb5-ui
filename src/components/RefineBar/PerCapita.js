import { GEO_NORM_NONE, GEO_NORM_PER1000 } from '../../constants';
import { useDispatch, useSelector } from 'react-redux';
import { dataNormalizationUpdated } from '../../reducers/filters/filtersSlice';
import { useMemo } from 'react';
import {
  selectFiltersDataNormalization,
  selectFiltersEnablePer1000,
} from '../../reducers/filters/selectors';
import { selectedClass } from '../../utils';

export const PerCapita = () => {
  const dataNormalization = useSelector(selectFiltersDataNormalization);
  const enablePer1000 = useSelector(selectFiltersEnablePer1000);
  const dispatch = useDispatch();

  const perCapButtonClass = useMemo(() => {
    if (enablePer1000) {
      return selectedClass(dataNormalization, GEO_NORM_PER1000);
    }
    return 'a-btn__disabled';
  }, [dataNormalization, enablePer1000]);

  return (
    <section className="per-capita">
      <p>Map shading</p>
      <div className="m-btn-group">
        <button
          aria-label="Display map by complaints"
          className={'a-btn' + selectedClass(dataNormalization, GEO_NORM_NONE)}
          onClick={() => {
            dispatch(dataNormalizationUpdated(GEO_NORM_NONE));
          }}
          disabled={dataNormalization === GEO_NORM_NONE}
        >
          Complaints
        </button>
        <button
          aria-label="Display map by complaints per 1,000 people"
          className={'a-btn ' + perCapButtonClass}
          disabled={dataNormalization === GEO_NORM_PER1000 || !enablePer1000}
          onClick={() => {
            dispatch(dataNormalizationUpdated(GEO_NORM_PER1000));
          }}
        >
          Complaints per 1,000 <span>population</span>
        </button>
      </div>
    </section>
  );
};
