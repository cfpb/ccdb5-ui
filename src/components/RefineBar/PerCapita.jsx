import { GEO_NORM_NONE, GEO_NORM_PER1000 } from '../../constants'
import { connect } from 'react-redux'
import { dataNormalizationChanged } from '../../actions/map';
import React from 'react'


export class PerCapita extends React.Component {
  _setNormalization( val ) {
    this.props.onDataNormalization( val )
  }

  _getRawButtonClass() {
    return this.props.dataNormalization === GEO_NORM_NONE ? 'selected' :
      'deselected'
  }

  _getPerCapButtonClass() {
    if ( this.props.enablePer1000 ) {
      return this.props.dataNormalization === GEO_NORM_PER1000 ? 'selected' :
        'deselected'
    }
    return 'a-btn__disabled'
  }
  render() {
    return (
      <section className="m-btn-group">
        <p>Map shading</p>
        <button
          className={ 'a-btn toggle-button raw ' + this._getRawButtonClass() }
          onClick={ () => this._setNormalization( GEO_NORM_NONE ) }>Complaints
        </button>
        <button
          className={ 'a-btn toggle-button capita ' + this._getPerCapButtonClass() }
          onClick={ () =>
              this.props.enablePer1000 &&
              this._setNormalization( GEO_NORM_PER1000 ) }>
          Complaints per 1,000 <span>population</span>
        </button>
      </section>
    )
  }
}

export const mapStateToProps = state => ( {
  dataNormalization: state.query.dataNormalization,
  enablePer1000: state.query.enablePer1000
} );

export const mapDispatchToProps = dispatch => ( {
  onDataNormalization: value => {
    dispatch( dataNormalizationChanged( value ) )
  }
} );

export default connect( mapStateToProps, mapDispatchToProps )( PerCapita )
