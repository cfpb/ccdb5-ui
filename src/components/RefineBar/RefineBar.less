@import (less) "../../css/base.less";

.refine-bar {
  section {
    margin: @grid_gutter-width/3 0 @grid_gutter-width/3 @grid_gutter-width/2;
    p {
      font-size: @size-vi;
      font-weight: 600;
      color: @gray;
    }
    button {
      &.selected {
        font-weight: 600;
      }
    }
    &:not(.filter-panel-toggle){
      button {
        color: @black;
      }
    }

    &.cf-select {
      border: none;
      select {
        background-color: @white;
        color: @black;
        border: 1px solid @gray-40;
      }
      &::after {
        background-color: @white;
        transform: translate(-5px, 0px);
      }
    }
  }

  .separator {
    @media @desktop, @tablet-lg, @phone {
      display: none;
    }
  }

  // Per Capita and Narratives buttons in the refine bar.
  .toggle-button {
    // Only used to hide "population" on per capita refine buttons.
    @media @phone, @tablet-sm {
      > span {
        display: none;
      }
    }

    &.selected {
      pointer-events: none;
    }

    &:not(.a-btn__disabled) {
      background-color: @pacific-80;
    }
    &.deselected {
      background-color: @pacific-40;
    }
  }
}
