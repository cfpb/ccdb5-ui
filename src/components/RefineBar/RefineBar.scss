@use 'sass:math';
@use '@cfpb/cfpb-design-system/src/abstracts' as *;
@use 'css/base' as *;
@use 'select' as *;

.refine-bar {
  section {
    margin: math.div($grid-gutter-width, 3) 0 math.div($grid-gutter-width, 3) math.div($grid-gutter-width, 2);
    p {
      font-size: $size-vi;
      font-weight: 600;
      color: var(--gray);
    }
    button {
      &.selected {
        background-color: var(--pacific-60);
        font-weight: 600;
      }
      &:not(.selected) {
        background-color: var(--pacific-20);
      }
    }
    &:not(.filter-panel-toggle) {
      button {
        color: var(--black);
      }
    }

    &.cf-select {
      border: none;
      select {
        background-color: var(--white);
        color: var(--black);
        border: 1px solid var(--gray-40);
      }
      &::after {
        background-color: var(--white);
        transform: translate(-5px, 0);
      }
    }
  }

  .separator {
    @media screen and (min-width: $layout-breakpoint-desktop-min) {
      display: none;
    }

    @media screen and (min-width: $layout-breakpoint-tablet-lg-min) and (max-width: $layout-breakpoint-tablet-lg-max) {
      display: none;
    }

    @media screen and (max-width: $layout-breakpoint-phone-max) {
      display: none;
    }
  }

  // Per Capita and Narratives buttons in the refine bar.
  button {
    // Only used to hide "population" on per capita refine buttons.
    @media screen and (max-width: $layout-breakpoint-phone-max),
      screen and (min-width: $layout-breakpoint-tablet-sm-min) and (max-width: $layout-breakpoint-tablet-sm-max){
      > span {
        display: none;
      }
    }

    &.disabled.a-btn__disabled {
      background-color: var(--gray-10);
    }
  }
}
