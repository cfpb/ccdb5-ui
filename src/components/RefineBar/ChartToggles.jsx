import './ChartToggles.scss';
import { chartTypeUpdated } from '../../reducers/trends/trendsSlice';
import getIcon from '../Common/Icon/iconMap';
import { selectedClass, sendAnalyticsEvent } from '../../utils';
import { useDispatch, useSelector } from 'react-redux';
import { selectTrendsChartType } from '../../reducers/trends/selectors';

export const ChartToggles = () => {
  const dispatch = useDispatch();
  const chartType = useSelector(selectTrendsChartType);

  const toggleChartType = (chartType) => {
    sendAnalyticsEvent('Button', 'Trends:' + chartType);
    dispatch(chartTypeUpdated(chartType));
  };

  return (
    <section className="chart-toggles">
      <p>Chart type</p>
      <div className="m-btn-group">
        <button
          aria-label="Toggle line chart"
          className={'a-btn' + selectedClass('line', chartType)}
          disabled={chartType === 'line'}
          onClick={() => {
            toggleChartType('line');
          }}
        >
          {getIcon('line-chart')}
        </button>
        <button
          aria-label="Toggle area chart"
          className={'a-btn' + selectedClass('area', chartType)}
          disabled={chartType === 'area'}
          onClick={() => {
            toggleChartType('area');
          }}
        >
          {getIcon('area-chart')}
        </button>
      </div>
    </section>
  );
};
