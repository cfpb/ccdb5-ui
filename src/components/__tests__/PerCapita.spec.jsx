import React from 'react';
import {mount} from 'enzyme';
import * as redux from 'react-redux';
import {PerCapita} from '../RefineBar/PerCapita';

describe('PerCapita', () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let mockDispatch;

  beforeEach(() => {
    // Mock useSelector hook
    spyOnUseSelector = jest.spyOn(redux, 'useSelector');
    // Mock useDispatch hook
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    // Mock dispatch function returned from useDispatch
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render', () => {
    spyOnUseSelector.mockReturnValue(false);
    const wrapper = mount(<PerCapita />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should dispatch dataNormalizationChanged GEO_NORM_NONE action', () => {
    spyOnUseSelector.mockReturnValue('Per 1000 pop.');
    const wrapper = mount(<PerCapita />);
    let button = wrapper.find('.capita');
    expect( button.hasClass( 'selected' ) ).toBeTruthy();

    button = wrapper.find('.raw');
    expect( button.hasClass( 'selected' ) ).toBeFalsy();

    button.simulate('click');
    expect(mockDispatch.mock.calls).toEqual([
      [{
        "type": "DATA_NORMALIZATION_SELECTED",
        "requery": "REQUERY_NEVER",
        "value": "None"
      }]
    ]);
  });

  it('should dispatch dataNormalizationChanged GEO_NORM_PER1000 action', () => {
    spyOnUseSelector.mockReturnValue('None');
    const wrapper = mount(<PerCapita />);
    let button = wrapper.find('.raw');
    expect( button.hasClass( 'selected' ) ).toBeTruthy();

    button = wrapper.find('.capita');
    expect( button.hasClass( 'selected' ) ).toBeFalsy();

    button.simulate('click');
    expect(mockDispatch.mock.calls).toEqual([
      [{
        "type": "DATA_NORMALIZATION_SELECTED",
        "requery": "REQUERY_NEVER",
        "value": "Per 1000 pop."
      }]
    ]);
  });

  it('should skip dataNormalizationChanged action when same', () => {
    spyOnUseSelector.mockReturnValue('None');
    const wrapper = mount(<PerCapita />);
    let button = wrapper.find('.raw');
    expect( button.hasClass( 'selected' ) ).toBeTruthy();

    button.simulate('click');
    expect(mockDispatch.mock.calls).toEqual([]);
  });

} );
