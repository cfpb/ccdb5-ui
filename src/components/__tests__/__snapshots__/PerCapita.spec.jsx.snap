// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: PerCapita initial state renders disabled enablePer1000 without crashing 1`] = `
<section
  className="m-btn-group"
>
  <p>
    Map shading
  </p>
  <button
    className="a-btn toggle-button raw selected"
    onClick={[Function]}
  >
    Complaints
  </button>
  <button
    className="a-btn toggle-button capita a-btn__disabled"
    onClick={[Function]}
  >
    Complaints per 1,000 
    <span>
      population
    </span>
  </button>
</section>
`;

exports[`component: PerCapita initial state renders enablePer1000 without crashing 1`] = `
<section
  className="m-btn-group"
>
  <p>
    Map shading
  </p>
  <button
    className="a-btn toggle-button raw selected"
    onClick={[Function]}
  >
    Complaints
  </button>
  <button
    className="a-btn toggle-button capita deselected"
    onClick={[Function]}
  >
    Complaints per 1,000 
    <span>
      population
    </span>
  </button>
</section>
`;
